New prescient-filter-method literal-prefix

tsdh opened the pull request:
The new `prescient-filter-method` value `literal-prefix` means the
first subquery must be the prefix of the candidate and the remaining
subqueries must be prefixes of words in the candidate.  Supports char
folding (just like `literal`).

As such, it is quite similar to `literal` except that it does strict
prefix matching instead of substring matching.

It is also similar to `prefix` except that it forces that the first
subquery matches at the beginning of the candidate.  Furthermore, it
only highlights the subqueries literally whereas the `prefix` method
highlights the complete matching words.

I'm happy to also update the `CHANGELOG.md` if this PR gets merged but
didn't do it right now so that different PRs don't provoke conflicts
in there.

----

# 

<!--

To expedite the pull request process, please see the contributor guide
for my projects:

  <https://github.com/raxod502/contributor-guide>

-->

raxod502 commented:
Sorry for the inconvenience, but the interface for filter methods was refactored in https://github.com/raxod502/prescient.el/pull/77 so that now each filter method is in a separate function instead of all of them being embedded in a case statement. Would you mind updating this PR accordingly? The actual changes look great though (pending readme and changelog mentions); thank you for your contribution!

tsdh commented:
@raxod502 I'll update the PR according to the changes in #77.

tsdh commented:
@raxod502 I've updated the PR and also added a commit adapting changelog and readme.

I had to change the signatures of all "make me a regexp for a
subquery" functions.  Now instead of `(query &optional with-groups)`
they take `(query props)` where `props` is a property list of the form
`(:with-groups <bool-value-or-all> :subquery-number <int-value>)`.  I
think that's the right approach when considering that probably there
will be more information needed to be passed to specific functions in
the future.  Adding more optional positional arguments which are
ignored by all but one functions didn't sound reasonable.

Of course, I've documented that in the docstring of the new alist variable `prescient-filter-alist` that has been introduced in #77.

raxod502 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,20 @@ The format is based on [Keep a Changelog].
+  but not `bar-foo` (because that doesn't start with `foo`) or
+  `foo-qbar` (because `bar` is no prefix of some word).
+  
+  It can be enabled bu adding `literal-prefix` to
```
```suggestion
  It can be enabled by adding `literal-prefix` to
```

raxod502 commented on prescient.el:
```diff
@@ -327,47 +347,65 @@ as a sub-query delimiter."
     regexp))
 
-(defun prescient-literal-regexp (query &optional with-groups)
+(defun prescient-literal-regexp (query props)
```
```suggestion
(cl-defun prescient-literal-regexp (query &key with-groups &allow-other-keys)
```

So it turns out that `cl-lib` has a neat way of doing this plist pattern out of the box. You just list the properties of the plist (without the leading colons), and they are automatically destructured if present, otherwise they default to nil. You have to include `&allow-other-keys` for forward compatibility.

In terms of the interface, this basically is the same as what you've already implemented, except that the `props` is a variadic argument (`&rest props`). In the docstring for `prescient-filter-alist`, we'd advise using the function signature with `cl-defun` and `&allow-other-keys`.

raxod502 commented on prescient.el:
```diff
@@ -327,47 +347,65 @@ as a sub-query delimiter."
+(defun prescient-literal-regexp (query props)
   "Return a regexp matching QUERY with character folding.
-If WITH-GROUPS is `all', enclose the match in a capture group."
+If property `:with-groups' in PROPS is `all', enclose the match
```
After using `cl-defun`, `with-groups` becomes a normal argument from the point of view of checkdoc.

```suggestion
If WITH-GROUPS is `all', enclose the match
```

raxod502 commented on prescient.el:
```diff
@@ -327,47 +347,65 @@ as a sub-query delimiter."
              "\\W*"))
 
-(defun prescient-regexp-regexp (query &optional _)
+(defun prescient-regexp-regexp (query _props)
```
This would look like so:

```suggestion
(defun prescient-regexp-regexp (query &rest _)
```

raxod502 reviewed (commented):
Well, this is technically a breaking change, but we only made these filter functions customizable at all very recently, so... YOLO. Seems fine to me. Just one suggestion to make this interface more ergonomic, and it all looks great otherwise!

tsdh commented:
@raxod502 If you like the `cl-defun` version more, feel free do do it that way.  IMHO, it adds more complexity for little value given that the alist is so new that this breaking change is not really breaking as you've said.

raxod502 commented:
The reason I'm suggesting `cl-defun` is not to preserve backwards compatibility. I don't think preserving such compatibility is important, as you point out. I just think using `cl-defun` will be simpler, since it reduces the amount of boilerplate code required.

tsdh commented:
@raxod502 i really don't see how `cl-defun` can be considered simpler.  I always have to look up its docs, especially for things like `&allow-other-keys`, most probably because it is seldomly used in elisp.  The only benefit I see is that the more complex signatures allow you to just use the parameter `with-groups` instead of `(plist-get props :with-groups)`.

Anyway, you don't have to convince me.  Just state your preference clearly and I'll change it accordingly if required.

tsdh commented:
@raxod502 Ok, I've refactored the changes to use `c-defun` with keyword arguments instead of passing just a plist.

raxod502 reviewed (approved):
Looks beautiful!

tsdh commented:
@raxod502 Wohoo!  Thanks a lot!
